x-pack/libbeat/management/generate.go

```diff
@@ -118,15 +118,15 @@ func deDotDataStream(raw dataStreamAndSource) (*proto.DataStream, error) {
 		return nil, fmt.Errorf("cannot unpack source field into struct: %w", err)
 	}
 
-	if ds.Dataset != "" && tmp.DataStream.Dataset != "" {
+	if ds.Dataset != "" && tmp.DataStream.Dataset != "" && (ds.Dataset != tmp.DataStream.Dataset) {
```

Comment:
The error message here isn't really right anymore, the key is supposed to be in both places. The only error is if they don't match.

Reply:
If they're the same it's not a problem, that's how the Elastic-Agent sends the data when the `data_stream` object is not flattened.

Reply:
The error is `duplicated key 'datastream.dataset` but the key being in both places actually the correct and expected result. The error now is only if they don't match, which the Elastic Agent should guarantee is impossible because the `DataStream` object is parsed out of the `Source` on the agent side but being defensive is fine.